week1/tictactoe/borboss366/Main.hs

```diff
@@ -0,0 +1,354 @@
+module Main where
+
+import Data.List (groupBy, maximumBy)
+import Data.Function(on)
+import Data.Matrix (matrix, Matrix, ncols, nrows, getElem, setElem)
+import Data.Char(toLower, isAsciiLower, ord, isDigit)
+import System.Console.ANSI (setSGR, ConsoleLayer(..), SGR(..), ColorIntensity(..), Color(..))
+import Text.Read(readMaybe)
+
+newtype TTTPoint = TTTPoint {
+  getPoint :: (Int, Int)
+} deriving (Eq, Show)
+
+newtype TTTDimension = TTTDimension {
+  getDimension :: (Int, Int)
+}  deriving (Eq, Show)
+
+newtype TTTLine = TTTLine {
+  getRow :: [TTTPoint]
+} deriving (Eq, Show)
+
+newtype TTTCellLine = TTTCellLine {
+  getCellRow :: [(TTTCellType, TTTPoint)]
+} deriving (Eq, Show)
+
+data TTTPlayer = Krestig | Nolek deriving (Show, Eq)
+data TTTCellType = Occupied TTTPlayer | EmptyCell deriving (Show, Eq)
+data TTTMoveState = Move TTTPlayer | Win TTTPlayer | TTTie deriving (Eq, Show)
+
+data TTTPosition = TTTPosition {
+  getField :: Matrix TTTCellType,
+  getMoves :: [TTTPoint],
+  get2Win :: Int
+} deriving (Eq, Show)
+
+startGame :: Int -> Int -> Int -> TTTPosition
+startGame w h r = TTTPosition {
+  getField = matrix h w (const EmptyCell),
+  getMoves = [],
+  get2Win = r
+}
+
+calcDimension :: TTTPosition -> TTTDimension
+calcDimension pos = TTTDimension (ncols field, nrows field)
+                    where field = getField pos
+
+calcSq :: TTTDimension -> Int
+calcSq (TTTDimension (w, h)) = w * h
+
+calcTotalPos :: TTTPosition -> Int
+calcTotalPos = calcSq . calcDimension
+
+calcPlayLength :: TTTPosition -> Int
+calcPlayLength = length . getMoves
+
+isOccupied :: TTTCellType -> Bool
+isOccupied (Occupied _) = True
+isOccupied _ = False
+
+isFinished :: TTTMoveState -> Bool
+isFinished (Win _) = True
+isFinished TTTie = True
+isFinished _ = False
+
+getRows :: TTTDimension -> [TTTLine]
+getRows (TTTDimension (w, h)) = [mkRow (0, y) (w - 1, y) | y <- [0..h - 1]]
+  where
+    mkRow p1 p2 = calcLineElems (TTTPoint p1) (TTTPoint p2)
+
+getLines :: TTTDimension -> [TTTLine]
```

Comment:
Comments would be useful here as well (e.g. to explain what's the difference between rows and lines).